pulsar/pulsar-impl/src/main/java/com/datastax/oss/sink/pulsar/AvroAPIAdapter.java

```diff
@@ -0,0 +1,221 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.sink.pulsar;
+
+import com.datastax.oss.sink.RetriableException;
+import com.datastax.oss.sink.config.ConfigException;
+import com.datastax.oss.sink.record.SchemaSupport;
+import com.google.common.collect.ImmutableMap;
+import java.nio.ByteBuffer;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericContainer;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.util.Utf8;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AvroAPIAdapter<Coat>
+    implements APIAdapter<
+        Coat, Object, LocalRecord<Coat, Object>, Schema, GenericRecord, Schema.Field, Header> {
+
+  public static final Logger log = LoggerFactory.getLogger(AvroAPIAdapter.class);
+
+  @Override
+  public RuntimeException adapt(ConfigException ex) {
+    return ex;
+  }
+
+  @Override
+  public RuntimeException adapt(RetriableException ex) {
+    return ex;
+  }
+
+  @Override
+  public String topic(LocalRecord<Coat, Object> record) {
+    return record.topic();
+  }
+
+  @Override
+  public Object key(LocalRecord<Coat, Object> record) {
+    return record.key();
+  }
+
+  @Override
+  public Object value(LocalRecord<Coat, Object> record) {
+    return record.payload();
+  }
+
+  @Override
+  public Long timestamp(LocalRecord<Coat, Object> record) {
+    return record.timestamp();
+  }
+
+  @Override
+  public Set<Header> headers(LocalRecord<Coat, Object> record) {
+    return record.headers();
+  }
+
+  @Override
+  public String headerKey(Header header) {
+    return header.name;
+  }
+
+  @Override
+  public Object headerValue(Header header) {
+    return header.value;
+  }
+
+  @Override
+  public Schema headerSchema(Header header) {
+    if (header.value == null) return schemaNull;
+    log.debug(
+        "get header {} {} {} {}",
+        header.name,
+        header.value,
+        header.value.getClass(),
+        primitiveSchemas.get(header.value.getClass()));
+    Schema schema = primitiveSchemas.get(header.value.getClass());
+    if (schema == null) schema = ((GenericContainer) header.value).getSchema();
+    return schema;
+  }
+
+  @Override
+  public boolean isStruct(Object object) {
+    return object instanceof GenericRecord;
+  }
+
+  @Override
+  public Schema schema(GenericRecord struct) {
+    return struct.getSchema();
+  }
+
+  @Override
+  public Set<String> fields(GenericRecord struct) {
```

Comment:
my local indent is 4-space. so, this is because I committed this file without fmt. (of course we should push changes after fmt, my fault. thanks for the note)